sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala

```diff
@@ -152,7 +168,7 @@ object NestedColumnAliasing {
     val exclusiveAttrSet = AttributeSet(exclusiveAttrs ++ otherRootReferences)
     val aliasSub = nestedFieldReferences.asInstanceOf[Seq[ExtractValue]]
       .filter(!_.references.subsetOf(exclusiveAttrSet))
-      .groupBy(_.references.head)
+      .groupBy(t => (t.references.head.exprId, t.references.head.dataType))
```

Comment:
Why need this change?

Reply:
https://github.com/apache/spark/pull/28898#discussion_r447950809

So basically the same expression can have different table aliases; thus cause collision during the map insertion. 

This is to group by the exprId, which is the unique identifier.

Reply:
I don't get it. It is grouping by attributes, why table aliases cause trouble here?

Reply:
Take a look at this query: 
```
with contact_rank as (
  select row_number() over (partition by address order by id desc) as rank,
  contacts.*
  from contacts
  order by name.last, name.first
)
select name.first, rank from contact_rank
```
```
nestedFieldReferences.asInstanceOf[Seq[ExtractValue]]
      .filter(!_.references.subsetOf(exclusiveAttrSet))
      .groupBy(_.references.head)
``` 
returns:
```
0 = {Tuple2@15534} "(name#46,List(name#46.first))"
1 = {Tuple2@15535} "(name#46,List(name#46.last, name#46.first))"
```

Basically `name#46` is the same attribute of type `AttributeReference`. The only attribute that is different is `qualifier`. Hence such change is needed.

Reply:
I see. I think we can simply clean up `qualifier` of returned `AttributeReference` in `collectRootReferenceAndExtractValue`.

We don't need the `qualifier`. We just use `exprId` and `dataType`.

Reply:
I think my method is cleaner because it only touches the `groupBy` logic by avoiding `exprId` key collision in the result map. I am not confident in changing `collectRootReferenceAndExtractValue`.

Reply:
We don't need to change `collectRootReferenceAndExtractValue`.

```scala
val aliasSub = nestedFieldReferences.asInstanceOf[Seq[ExtractValue]]
  .filter(!_.references.subsetOf(exclusiveAttrSet))
  .groupBy(_.references.head.withQualifier(Seq.empty))
  ...
```

Reply:
Yeah, but my solution solves the problem in general. Let's say it fails for a different reason that causes `exprId` collision. My `groupBy` ensures that collision isn't ever possible.

Reply:
I don't think `exprId` will has collision. `exprId` should be unique. If there is collision, the query plan should have some place wrong...

Reply:
Ok. Sorry for the confusion. I meant this [line](https://github.com/apache/spark/blob/d043e3846432981ac7f9b85cb4515499ab6f0118/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala#L207). My solution guarantee that in the event of something like `exprIdA->List(a,b)` and `exprIdA->List(c,d)`, my `groupBy` will always do the correct thing, no matter the actual cause (it might not be `qualifier` issue in other cases)

Reply:
Ok. Can you add a comment here?

Reply:
Sure.

Reply:
Just added.